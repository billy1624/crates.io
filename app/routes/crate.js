import { NotFoundError } from '@ember-data/adapter/error';
import Route from '@ember/routing/route';
import { inject as service } from '@ember/service';
import { ajax_fail, htmlDecode } from '../utils/ajax';

export default class CrateRoute extends Route {
  @service headData;
  @service router;
  @service store;

  async model(params, transition) {
    let crateName = params.crate_id;

    try {
      return (async () => {
        const data = await ajax_fail(
          `https://raw.githubusercontent.com/SeaQL/rustacean.info/main/related-articles/${crateName}.json`,
        );
        const res = await this.store.findRecord('crate', crateName);
        Object.assign(res, {
          articles: data.map(row => {
            row.title = htmlDecode(row.title);
            return row;
          }),
          link_contribute_articles: `https://github.com/SeaQL/rustacean.info/blob/main/CONTRIBUTING.md`,
          link_create_articles: `https://github.com/SeaQL/rustacean.info/new/main/related-articles`,
          link_edit_articles: `https://github.com/SeaQL/rustacean.info/edit/main/related-articles/${crateName}.json`,
        });
        return res;
      })();
    } catch (error) {
      if (error instanceof NotFoundError) {
        let title = `${crateName}: Crate not found`;
        this.router.replaceWith('catch-all', { transition, error, title });
      } else {
        let title = `${crateName}: Failed to load crate data`;
        this.router.replaceWith('catch-all', { transition, error, title, tryAgain: true });
      }
    }
  }

  setupController(controller, model) {
    super.setupController(...arguments);
    this.headData.crate = model;
  }

  resetController() {
    super.resetController(...arguments);
    this.headData.crate = null;
  }
}
