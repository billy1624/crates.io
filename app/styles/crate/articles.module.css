.results-meta {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: var(--space-2xs);

    @media only screen and (max-width: 550px) {
        display: block;
    }
}

.page-description {
    composes: small from '../shared/typography.module.css';

    @media only screen and (max-width: 550px) {
        display: block;
        margin-bottom: var(--space-s);
    }
}
.list {
    list-style: none;
    margin: 0 0 var(--space-s);
    padding: 0;

    > * + * {
        margin-top: var(--space-2xs);
    }
}
