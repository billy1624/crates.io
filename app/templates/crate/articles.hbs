<CrateHeader @crate={{this.model}} />

<ul local-class="list">
  {{#if this.model.articles}}
    {{#each this.model.articles as |article|}}
      <li>
        <ArticleList::Row @link={{article.link}} @title={{article.title}} local-class="row" data-test-version={{article.title}} />
      </li>
    {{/each}}
    <br>
    Contribute and post more related articles on <a href={{this.model.link_edit_articles}} target="_blank">GitHub</a>!
  {{else}}
    Contribute and post the first related article on <a href={{this.model.link_create_articles}} target="_blank">GitHub</a>!
  {{/if}}
  Read contributing guide <a href={{this.model.link_contribute_articles}} target="_blank">here</a>.
</ul>
