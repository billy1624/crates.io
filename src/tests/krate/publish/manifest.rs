use crate::builders::PublishBuilder;
use crate::util::insta::{any_id_redaction, id_redaction};
use crate::util::{RequestHelper, TestApp};
use http::StatusCode;
use insta::assert_json_snapshot;

#[tokio::test(flavor = "multi_thread")]
async fn boolean_readme() {
    // see https://github.com/rust-lang/crates.io/issues/6847

    let (_app, _anon, _cookie, token) = TestApp::full().with_token();

    let response = token
        .publish_crate(PublishBuilder::new("foo", "1.0.0").custom_manifest(
            r#"[package]
            name = "foo"
            version = "1.0.0"
            description = "description"
            license = "MIT"
            rust-version = "1.69"
            readme = false"#,
        ))
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".crate.created_at" => "[datetime]",
        ".crate.updated_at" => "[datetime]",
    });

    let response = token.get::<()>("https://crates.io/api/v1/crates/foo/1.0.0").await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".version.id" => any_id_redaction(),
        ".version.created_at" => "[datetime]",
        ".version.updated_at" => "[datetime]",
        ".version.published_by.id" => id_redaction(token.as_model().user_id),
        ".version.audit_actions[].time" => "[datetime]",
        ".version.audit_actions[].user.id" => id_redaction(token.as_model().user_id),
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn missing_manifest() {
    let (_app, _anon, _cookie, token) = TestApp::full().with_token();

    let response = token
        .publish_crate(PublishBuilder::new("foo", "1.0.0").no_manifest())
        .await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_json_snapshot!(response.json());
}

#[tokio::test(flavor = "multi_thread")]
async fn manifest_casing() {
    let (_app, _anon, _cookie, token) = TestApp::full().with_token();

    let response = token
        .publish_crate(
            PublishBuilder::new("foo", "1.0.0")
                .add_file(
                    "foo-1.0.0/CARGO.TOML",
                    "[package]\nname = \"foo\"\nversion = \"1.0.0\"\n",
                )
                .no_manifest(),
        )
        .await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_json_snapshot!(response.json());
}

#[tokio::test(flavor = "multi_thread")]
async fn multiple_manifests() {
    let (_app, _anon, _cookie, token) = TestApp::full().with_token();

    let response = token
        .publish_crate(
            PublishBuilder::new("foo", "1.0.0")
                .add_file(
                    "foo-1.0.0/Cargo.toml",
                    "[package]\nname = \"foo\"\nversion = \"1.0.0\"\n",
                )
                .add_file(
                    "foo-1.0.0/cargo.toml",
                    "[package]\nname = \"foo\"\nversion = \"1.0.0\"\n",
                )
                .no_manifest(),
        )
        .await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_json_snapshot!(response.json());
}

#[tokio::test(flavor = "multi_thread")]
async fn invalid_manifest() {
    let (_app, _anon, _cookie, token) = TestApp::full().with_token();

    let response = token
        .publish_crate(PublishBuilder::new("foo", "1.0.0").custom_manifest(""))
        .await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_json_snapshot!(response.json());
}

#[tokio::test(flavor = "multi_thread")]
async fn invalid_manifest_missing_name() {
    let (_app, _anon, _cookie, token) = TestApp::full().with_token();

    let response = token
        .publish_crate(
            PublishBuilder::new("foo", "1.0.0").custom_manifest("[package]\nversion = \"1.0.0\""),
        )
        .await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_json_snapshot!(response.json());
}

#[tokio::test(flavor = "multi_thread")]
async fn invalid_manifest_missing_version() {
    let (_app, _anon, _cookie, token) = TestApp::full().with_token();

    let response = token
        .publish_crate(
            PublishBuilder::new("foo", "1.0.0").custom_manifest("[package]\nname = \"foo\""),
        )
        .await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_json_snapshot!(response.json());
}

#[tokio::test(flavor = "multi_thread")]
async fn invalid_rust_version() {
    let (_app, _anon, _cookie, token) = TestApp::full().with_token();

    let response =
        token.publish_crate(PublishBuilder::new("foo", "1.0.0").custom_manifest(
            "[package]\nname = \"foo\"\nversion = \"1.0.0\"\ndescription = \"description\"\nlicense = \"MIT\"\nrust-version = \"\"\n",
        )).await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_json_snapshot!(response.json());

    let response = token.publish_crate(PublishBuilder::new("foo", "1.0.0").custom_manifest(
        "[package]\nname = \"foo\"\nversion = \"1.0.0\"\ndescription = \"description\"\nlicense = \"MIT\"\nrust-version = \"1.0.0-beta.2\"\n",
    )).await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_json_snapshot!(response.json());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_lib_and_bin_crate() {
    let (_app, _anon, _cookie, token) = TestApp::full().with_token();

    let publish_builder = PublishBuilder::new("foo", "1.0.0")
        .add_file("foo-1.0.0/src/lib.rs", "pub fn foo() {}")
        .add_file("foo-1.0.0/src/main.rs", "fn main() {}")
        .add_file("foo-1.0.0/src/bin/bar.rs", "fn main() {}");

    let response = token.publish_crate(publish_builder).await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".crate.created_at" => "[datetime]",
        ".crate.updated_at" => "[datetime]",
    });

    let response = token.get::<()>("https://crates.io/api/v1/crates/foo/1.0.0").await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".version.id" => any_id_redaction(),
        ".version.created_at" => "[datetime]",
        ".version.updated_at" => "[datetime]",
        ".version.published_by.id" => id_redaction(token.as_model().user_id),
        ".version.audit_actions[].time" => "[datetime]",
        ".version.audit_actions[].user.id" => id_redaction(token.as_model().user_id),
    });
}
