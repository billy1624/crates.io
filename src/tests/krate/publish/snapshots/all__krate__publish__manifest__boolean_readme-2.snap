---
source: src/tests/krate/publish/manifest.rs
expression: response.json()
---
{
  "version": {
    "audit_actions": [
      {
        "action": "publish",
        "time": "[datetime]",
        "user": {
          "avatar": null,
          "id": "[id]",
          "login": "foo",
          "name": null,
          "url": "https://github.com/foo"
        }
      }
    ],
    "bin_names": [],
    "checksum": "418c24c55cbcde274a5b1815d2c844203c7116e7f712008f4727d38bc7c82d8d",
    "crate": "foo",
    "crate_size": 170,
    "created_at": "[datetime]",
    "dl_path": "https://crates.io/api/v1/crates/foo/1.0.0/download",
    "downloads": 0,
    "features": {},
    "has_lib": false,
    "id": "[id]",
    "lib_links": null,
    "license": "MIT",
    "links": {
      "authors": "https://crates.io/api/v1/crates/foo/1.0.0/authors",
      "dependencies": "https://crates.io/api/v1/crates/foo/1.0.0/dependencies",
      "version_downloads": "https://crates.io/api/v1/crates/foo/1.0.0/downloads"
    },
    "num": "1.0.0",
    "published_by": {
      "avatar": null,
      "id": "[id]",
      "login": "foo",
      "name": null,
      "url": "https://github.com/foo"
    },
    "readme_path": "https://crates.io/api/v1/crates/foo/1.0.0/readme",
    "rust_version": "1.69",
    "updated_at": "[datetime]",
    "yanked": false
  }
}
